src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -58,16 +58,140 @@ private enum ImageFileMachine
         private IntPtr _unmanagedCallbacks; // array of pointers to JIT-EE interface callbacks
         private Object _keepAlive; // Keeps delegates for the callbacks alive
 
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        private delegate IntPtr JitStartupDelegate(IntPtr host);
+
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        private delegate IntPtr GetJitDelegate();
+
+        private delegate IntPtr GetJitHostDelegate(IntPtr configProvider);
+        private delegate CorJitResult JitCompileMethodDelegate(out IntPtr exception,
+            IntPtr jit, IntPtr thisHandle, IntPtr callbacks,
+            ref CORINFO_METHOD_INFO info, uint flags, out IntPtr nativeEntry, out uint codeSize);
+        private delegate IntPtr AllocExceptionDelegate([MarshalAs(UnmanagedType.LPWStr)]string message, int messageLength);
+        private delegate void FreeExceptionDelegate(IntPtr obj);
+        private delegate char* GetExceptionMessageDelegate(IntPtr obj);
+
+        private static IntPtr s_jitLibrary;
+        private static IntPtr s_jitHostLibrary;
+
+        private static JitStartupDelegate s_jitStartup;
+        private static GetJitDelegate s_getJit;
+
+        private static GetJitHostDelegate s_getJitHost;
+        private static JitCompileMethodDelegate s_jitCompileMethod;
+        private static AllocExceptionDelegate s_allocException;
+        private static FreeExceptionDelegate s_freeException;
+        private static GetExceptionMessageDelegate s_getExceptionMessage;
+
         private ExceptionDispatchInfo _lastException;
 
-        [DllImport(JitLibrary, CallingConvention = CallingConvention.StdCall)] // stdcall in CoreCLR!
-        private extern static IntPtr jitStartup(IntPtr host);
+        public static void Startup(TargetOS targetOS, TargetArchitecture targetArchitecture)
+        {
+            Debug.Assert(s_jitLibrary == IntPtr.Zero);
+            Debug.Assert(s_jitHostLibrary == IntPtr.Zero);
+
+            string crossgenPath = Path.GetDirectoryName(typeof(CorInfoImpl).Assembly.Location);
+            string jitHostPath = Path.Combine(crossgenPath, GetHostSpec());
+            string jitHostLibraryName = Path.Combine(jitHostPath, GetLibraryPrefix() + JitSupportLibrary + GetLibraryExtension());
```

Comment:
Without switching jitinterface we would need to publish two versions of Crossgen2 for arm64, one cross-compiling (x64-hosted) for compiling the framework on the x64 build machine and another native arm64-hosted jitinterface used when running Crossgen2 on the ARM64 Helix HW to compile the test apps.

Reply:
In this sense the jitinterface is dependent on the host whereas clrjit is dependent on both the host and target, that's the gist of this change. I believe it smoothly translates to supplying additional variants of jitinterface / jit to achieve all cross-targeting we need.

Reply:
For instance, we can add x86 support tomorrow by just publishing x86 jitinterface and clrjit along with Crossgen2 to the proper folders and fixing 2-3 bugs.

Reply:
And at some point we can end up merging them all by just creating a union of the subfolders and getting a fully cross-targeting Crossgen2 package.

Reply:
Hmm, not sure about x86, we'd probably need Crossgen2 with the 32-bit set but anyway :-).

Reply:
@trylek that only makes sense if the crossgen2 compiler package we produce is platform agnostic. I would expect us to configure the world such that the x64 crossgen2 can target arm64 and x64.

Reply:
Yup, what David says - crossgen2 is not an AnyCpu project - it's compiled for a specific host OS/arch. This host OS/arch will match the host OS/arch of the SDK itself, the host OS/arch of JitSupport, and the host OS/arch of clrjit.

We need a probing logic for clrjit, because because we have multiple clrjits per __target__ (but still there is just one __host__).

JitSupport is __target__ agnostic.

Reply:
OK, I treat this as a preliminary consensus that CoreCLR arm64 build should end up building two Crossgen2 packages - one cross-targeting under <code>artifacts\bin\coreclr\Windows_NT.arm64.Checked\x64\crossgen2</code> that's going to be used for cross-compilation of the framework assemblies on the build machines and a native arm64-hosted under <code>artifacts\bin\coreclr\Windows_NT.arm64.Checked\crossgen2</code> and <code>CORE_ROOT\crossgen2</code> that's going to be subsequently used for native arm64 test compilation on the Helix arm64 machines. I'm going to modify (simplify) the change based on your suggestion.